queasars/minimum_eigensolvers/evqe/evolutionary_algorithm/individual.py

```diff
@@ -0,0 +1,736 @@
+        )
+
+    @staticmethod
+    def remove_layers(individual: "EVQEIndividual", n_layers: int) -> "EVQEIndividual":
+        """Returns a new individual, based on the given individual, but with the last n_layers removed
+
+        :arg individual: on which the new individual is based on
+        :type individual: EVQEIndividual
+        :arg n_layers: amount of last layers to remove
+        :type n_layers: int
+        :return: the new individual
+        :rtype: EVQEIndividual
+        """
+
+        if not 0 < n_layers:
+            raise EVQEIndividualException("n_layers must be at least 1!")
+        if not n_layers < len(individual.layers):
+            raise EVQEIndividualException(
+                "Layer removal must leave at least one layer remaining! Choose a smaller n_layer value"
+            )
+
+        # Remove the last layers
+        layers: list[EVQECircuitLayer] = list(individual.layers)[0 : len(individual.layers) - n_layers]
+        # Get the parameter values for the remaining layers
+        parameter_values: list[float] = list(individual.parameter_values)[
+            0 : individual.layer_parameter_indices[len(individual.layers) - n_layers][0]
+        ]
+
+        return EVQEIndividual(
+            n_qubits=individual.n_qubits,
+            layers=tuple(layers),
+            parameter_values=tuple(parameter_values),
+        )
+
+    @staticmethod
+    def get_genetic_distance(individual_1: "EVQEIndividual", individual_2: "EVQEIndividual") -> int:
+        """
+        Returns the genetic distance between two individuals, which is defined as the sum of the amount of
+        all layers of both individuals subtracted by the amount of shared layers
+
+        :param individual_1: individual to determine genetic distance from
+        :type individual_1: EVQEIndividual
+        :param individual_2: individual to determine genetic distance from
+        :type individual_2: EVQEIndividual
+        :return:
+        """
+        n_all_layers: int = ceil(0.5 * (len(individual_1.layers) + len(individual_2.layers)))
+        n_shared_layers: int = len({layer for layer in individual_1.layers if layer in individual_2.layers})
+
+        return n_all_layers - n_shared_layers
+
+    def __post_init__(self) -> None:
+        # Disallow the initialization of invalid individuals
+        if not self.is_valid():
+            raise EVQEIndividualException("The created individual is not valid!")
+
+        # Initialize an immutable mapping to hold the parameter indices for each layer
+        layer_parameter_indices: dict[int, tuple[int, ...]] = {}
+        parameter_index: int = 0
+        for i, layer in enumerate(self.layers):
+            layer_parameter_indices[i] = tuple(range(parameter_index, parameter_index + layer.n_parameters))
+            parameter_index += layer.n_parameters
+        object.__setattr__(self, "_layer_parameter_indices", MappingProxyType(layer_parameter_indices))
```

Comment:
Is there an actual way to create an immutable dictionary?

Reply:
As discussed in private chat, there is no real immutable dictionary in standard python. So i don't think there is another way of handling this? But since you are correct and the dictionary is technically not truly immutable, i adjusted the description in the docstring accordingly.

Reply:
Yeah, sorry about that! In Scala, there is a whole separate namespace for mutable and immutalbe collections and I have been bitten by this before in Python.